src/queue.rs

```diff
@@ -83,59 +83,71 @@ impl<T: Send> Idle<T> {
 /// (including those which are not in the queue.)
 #[derive(Debug)]
 pub struct Queue<C: Send> {
-    idle: SegQueue<Idle<C>>,
+    idle: ArrayQueue<Idle<C>>,
     total_count: AtomicUsize,
+    idle_push_error_count: AtomicUsize,
 }
 
 impl<C: Send> Queue<C> {
     /// Construct an empty queue with a certain capacity
-    pub fn new() -> Queue<C> {
+    pub(crate) fn new(capacity: usize) -> Queue<C> {
         Queue {
-            idle: SegQueue::new(),
+            idle: ArrayQueue::new(capacity),
             total_count: AtomicUsize::new(0),
+            idle_push_error_count: AtomicUsize::new(0),
         }
     }
 
     /// Count of idle connection in queue
     #[inline(always)]
-    pub fn idle(&self) -> usize {
+    pub(crate) fn idle(&self) -> usize {
         self.idle.len()
     }
 
     /// Count of total connections active
     #[inline(always)]
-    pub fn total(&self) -> usize {
+    pub(crate) fn total(&self) -> usize {
         self.total_count.load(Ordering::SeqCst)
     }
 
+    /// Error counter incremented in `store()` method above.
+    pub(crate) fn idle_push_error_count(&self) -> usize {
+        self.idle_push_error_count.load(Ordering::Relaxed)
+    }
+
     /// Push a new connection into the queue (this will increment
     /// the total connection count).
-    pub fn new_conn(&self, conn: Live<C>) {
-        self.store(conn);
-        self.increment();
+    pub(crate) fn new_conn(&self, conn: Live<C>) {
+        if self.store(conn).is_ok() {
+            self.increment();
+        }
     }
 
     /// Store a connection which has already been counted in the queue
     /// (this will NOT increment the total connection count).
-    pub fn store(&self, conn: Live<C>) {
-        self.idle.push(Idle::new(conn));
+    pub(crate) fn store(&self, conn: Live<C>) -> Result<(), PushError<Idle<C>>> {
```

Comment:
I can't think of a time when this would return an error, as we only put a new connection into the pool after removing an old connection, but I guess it's better to be safe

Reply:
I feel the same and am even thinking to make the size of `ArrayQueue` to the maximum connections, e.g. 64, but just to be safe.  And since the error count is exposed through the API, we can monitor the behavior and reduce it to max count in the future.  Thanks for the keen look, Nate, as usual.